netket/models/autoreg.py

```diff
@@ -149,18 +166,34 @@ def setup(self):
             for i in range(self.layers)
         ]
 
-    def conditionals(self, inputs: Array, cache: PyTree) -> Tuple[Array, PyTree]:
-        return _conditionals(self, inputs, cache)
+    def conditional(self, inputs: Array, index: int) -> Array:
+        return _conditionals(self, inputs)[:, index, :]
+
+    def conditionals(self, inputs: Array) -> Array:
+        return _conditionals(self, inputs)
 
     def __call__(self, inputs: Array) -> Array:
         return _call(self, inputs)
 
 
-def _conditional_log_psi(
-    model: ARNN, inputs: Array, cache: PyTree
-) -> Tuple[Array, PyTree]:
+def l2_normalize(log_psi: Array, eps: float) -> Array:
+    """
+    Normalizes log_psi to have L2-norm 1 along the last axis.
+
+    Args:
+      eps: a small number for better numerical stability in log.
     """
-    Computes the log of the conditional wave-functions for each spin if it takes each value.
+    log_psi_sqr = 2 * log_psi.real
+    max_log_psi_sqr = log_psi_sqr.max(axis=-1, keepdims=True)
+    # Subtract max_log_psi_sqr for better numerical stability
+    sum_psi_sqr = jnp.exp(log_psi_sqr - max_log_psi_sqr).sum(axis=-1, keepdims=True)
+    log_psi -= 1 / 2 * (jnp.log(sum_psi_sqr + eps) + max_log_psi_sqr)
```

Comment:
why don't you use ```log_sum_exp``` directly instead of doing it by hand ?

Reply:
Ok, I just realized that it's equivalent to a `logsumexp`